File: Pinball App 2/Pinball App 2/library/LibrarySelectionSupport.swift
import Foundation

struct PinballLibrarySelectionResolution {
    let selectedSourceID: String
    let sortOption: PinballLibrarySortOption
    let yearSortDescending: Bool
    let selectedBank: Int?
}

func resolvePreferredLibrarySource(
    sources: [PinballLibrarySource],
    selectedSourceID: String?,
    currentSelectedSourceID: String? = nil
) -> PinballLibrarySource? {
    let preferredSourceID = [selectedSourceID, currentSelectedSourceID]
        .compactMap { canonicalLibrarySourceID($0) }
        .first(where: { id in sources.contains(where: { $0.id == id }) })
    return sources.first(where: { $0.id == preferredSourceID }) ?? sources.first
}

func resolveLibrarySelection(
    payload: PinballLibraryPayload,
    sourceState: PinballLibrarySourceState,
    currentSelectedSourceID: String
) -> PinballLibrarySelectionResolution? {
    let chosenSource = resolvePreferredLibrarySource(
        sources: payload.sources,
        selectedSourceID: sourceState.selectedSourceID,
        currentSelectedSourceID: currentSelectedSourceID
    )
    return chosenSource.map { source in
        resolveLibrarySelectionForSource(
            source: source,
            games: payload.games,
            sourceState: sourceState
        )
    }
}

func resolveLibrarySelectionForSource(
    source: PinballLibrarySource,
    games: [PinballGame],
    sourceState: PinballLibrarySourceState
) -> PinballLibrarySelectionResolution {
    let sourceGames = games.filter { $0.sourceId == source.id }
    let options = librarySortOptions(for: source, games: sourceGames)
    let persistedSort = sourceState.selectedSortBySource[source.id]

    let selection: (sortOption: PinballLibrarySortOption, yearSortDescending: Bool)
    if source.type == .manufacturer {
        selection = (.year, true)
    } else if persistedSort == "YEAR_DESC", options.contains(.year) {
        selection = (.year, true)
    } else if let persistedSort,
              let sortOption = PinballLibrarySortOption(rawValue: persistedSort),
              options.contains(sortOption) {
        selection = (sortOption, false)
    } else {
        let defaultSort = libraryPreferredDefaultSortOption(for: source, games: sourceGames)
        let resolvedSort = options.contains(defaultSort) ? defaultSort : (options.first ?? .alphabetical)
        selection = (
            resolvedSort,
            resolvedSort == .year ? libraryPreferredDefaultYearSortDescending(for: source, games: sourceGames) : false
        )
    }

    let selectedBank: Int?
    if source.type == .venue, sourceGames.contains(where: { ($0.bank ?? 0) > 0 }) {
        selectedBank = sourceState.selectedBankBySource[source.id]
    } else {
        selectedBank = nil
    }

    return PinballLibrarySelectionResolution(
        selectedSourceID: source.id,
        sortOption: selection.sortOption,
        yearSortDescending: selection.yearSortDescending,
        selectedBank: selectedBank
    )
}

func librarySortOptions(
    for source: PinballLibrarySource,
    games: [PinballGame]
) -> [PinballLibrarySortOption] {
    switch source.type {
    case .category, .manufacturer, .tournament:
        return [.year, .alphabetical]
    case .venue:
        let hasBank = games.contains { ($0.bank ?? 0) > 0 }
        var options: [PinballLibrarySortOption] = [.area]
        if hasBank { options.append(.bank) }
        options.append(.alphabetical)
        options.append(.year)
        return options
    }
}

func libraryPreferredDefaultSortOption(
    for source: PinballLibrarySource,
    games: [PinballGame]
) -> PinballLibrarySortOption {
    switch source.type {
    case .manufacturer:
        return .year
    case .category, .tournament:
        return .alphabetical
    case .venue:
        let hasArea = games.contains {
            guard let area = $0.area?.trimmingCharacters(in: .whitespacesAndNewlines) else { return false }
            return !area.isEmpty && area.lowercased() != "null"
        }
        let hasPosition = games.contains { ($0.group ?? 0) > 0 || ($0.pos ?? 0) > 0 }
        return (hasArea || hasPosition) ? .area : .alphabetical
    }
}

func libraryPreferredDefaultYearSortDescending(
    for source: PinballLibrarySource,
    games: [PinballGame]
) -> Bool {
    source.type == .manufacturer && libraryPreferredDefaultSortOption(for: source, games: games) == .year
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibrarySelectionSupport.kt
package com.pillyliu.pinprofandroid.library

internal data class LibrarySelectionResolution(
    val selectedSourceId: String,
    val sortOptionName: String,
    val yearSortDescending: Boolean,
    val selectedBank: Int?,
)

internal fun resolvePreferredLibrarySource(
    sources: List<LibrarySource>,
    selectedSourceId: String?,
    currentSelectedSourceId: String? = null,
): LibrarySource? {
    val preferredSourceId = listOfNotNull(selectedSourceId, currentSelectedSourceId)
        .mapNotNull(::canonicalLibrarySourceId)
        .firstOrNull { candidate -> sources.any { it.id == candidate } }
    return sources.firstOrNull { it.id == preferredSourceId } ?: sources.firstOrNull()
}

internal fun resolveLibrarySelection(
    payload: ParsedLibraryData,
    sourceState: LibrarySourceState,
    currentSelectedSourceId: String,
): LibrarySelectionResolution? {
    val chosenSource = resolvePreferredLibrarySource(
        sources = payload.sources,
        selectedSourceId = sourceState.selectedSourceId,
        currentSelectedSourceId = currentSelectedSourceId,
    )
    return chosenSource?.let { source ->
        resolveLibrarySelectionForSource(
            source = source,
            games = payload.games,
            sourceState = sourceState,
        )
    }
}

internal fun resolveLibrarySelectionForSource(
    source: LibrarySource,
    games: List<PinballGame>,
    sourceState: LibrarySourceState,
): LibrarySelectionResolution {
    val sourceGames = games.filter { it.sourceId == source.id }
    val options = sortOptionsForSource(source, sourceGames)
    val persistedSort = sourceState.selectedSortBySource[source.id]
    val (sortOptionName, yearSortDescending) = when {
        source.type == LibrarySourceType.MANUFACTURER ->
            LibrarySortOption.YEAR.name to true
        persistedSort == "YEAR_DESC" ->
            LibrarySortOption.YEAR.name to true
        persistedSort != null && options.any { it.name == persistedSort } ->
            persistedSort to false
        else -> {
            val defaultSort = preferredDefaultSortOption(source, sourceGames)
            val resolvedSort = defaultSort.takeIf { options.contains(it) } ?: options.first()
            resolvedSort.name to preferredDefaultYearSortDescending(source, sourceGames)
        }
    }
    val selectedBank = if (source.type == LibrarySourceType.VENUE && sourceGames.any { (it.bank ?: 0) > 0 }) {
        sourceState.selectedBankBySource[source.id]
    } else {
        null
    }
    return LibrarySelectionResolution(
        selectedSourceId = source.id,
        sortOptionName = sortOptionName,
        yearSortDescending = yearSortDescending,
        selectedBank = selectedBank,
    )
}

internal fun preferredDefaultSortOption(source: LibrarySource, games: List<PinballGame>): LibrarySortOption {
    return when (source.type) {
        LibrarySourceType.MANUFACTURER -> LibrarySortOption.YEAR
        LibrarySourceType.CATEGORY,
        LibrarySourceType.TOURNAMENT -> LibrarySortOption.ALPHABETICAL
        LibrarySourceType.VENUE -> {
            val hasArea = games.any {
                val area = it.area?.trim()
                !area.isNullOrEmpty() && !area.equals("null", ignoreCase = true)
            }
            val hasPosition = games.any { (it.group ?: 0) > 0 || (it.position ?: 0) > 0 }
            if (hasArea || hasPosition) LibrarySortOption.AREA else LibrarySortOption.ALPHABETICAL
        }
    }
}

internal fun preferredDefaultYearSortDescending(source: LibrarySource, games: List<PinballGame>): Boolean =
    source.type == LibrarySourceType.MANUFACTURER && preferredDefaultSortOption(source, games) == LibrarySortOption.YEAR
